pkg/apis/resource/validation/validation.go

```diff
@@ -433,179 +512,244 @@ func validateNodeName(name string, fldPath *field.Path) field.ErrorList {
+func validateDevice(device resource.Device, fldPath *field.Path) field.ErrorList {
 	var allErrs field.ErrorList
-	driverNames := sets.New[string]()
-	for i, request := range requests {
-		idxPath := fldPath.Index(i)
-		driverName := request.DriverName
-		allErrs = append(allErrs, validateResourceDriverName(driverName, idxPath.Child("driverName"))...)
-		if driverNames.Has(driverName) {
-			allErrs = append(allErrs, field.Duplicate(idxPath.Child("driverName"), driverName))
-		} else {
-			driverNames.Insert(driverName)
-		}
-		allErrs = append(allErrs, validateResourceRequests(request.Requests, idxPath.Child("requests"), requestStored)...)
+	allErrs = append(allErrs, validateDeviceName(device.Name, fldPath.Child("name"))...)
+	if device.Basic == nil {
+		allErrs = append(allErrs, field.Required(fldPath.Child("basic"), ""))
+	} else {
+		allErrs = append(allErrs, validateBasicDevice(*device.Basic, fldPath.Child("basic"))...)
 	}
 	return allErrs
 }
 
-func validateResourceRequests(requests []resource.ResourceRequest, fldPath *field.Path, requestStored bool) field.ErrorList {
+func validateBasicDevice(device resource.BasicDevice, fldPath *field.Path) field.ErrorList {
 	var allErrs field.ErrorList
-	for i, request := range requests {
-		idxPath := fldPath.Index(i)
-		allErrs = append(allErrs, validateResourceRequestModel(&request.ResourceRequestModel, idxPath, requestStored)...)
-	}
-	if len(requests) == 0 {
-		// We could allow this, it just doesn't make sense: the entire entry would get ignored and thus
-		// should have been left out entirely.
-		allErrs = append(allErrs, field.Required(fldPath, "empty entries with no requests are not allowed"))
+	allErrs = append(allErrs, validateMap(device.Attributes, resource.ResourceSliceMaxAttributesAndCapacitiesPerDevice, validateQualifiedName, validateDeviceAttribute, fldPath.Child("attributes"))...)
+	allErrs = append(allErrs, validateMap(device.Capacity, resource.ResourceSliceMaxAttributesAndCapacitiesPerDevice, validateQualifiedName, validateQuantity, fldPath.Child("capacity"))...)
+	if combinedLen, max := len(device.Attributes)+len(device.Capacity), resource.ResourceSliceMaxAttributesAndCapacitiesPerDevice; combinedLen > max {
+		allErrs = append(allErrs, field.Invalid(fldPath, combinedLen, fmt.Sprintf("the total number of attributes and capacities must not exceed %d", max)))
 	}
 	return allErrs
 }
 
-func validateResourceRequestModel(model *resource.ResourceRequestModel, fldPath *field.Path, requestStored bool) field.ErrorList {
+var (
+	numericIdentifier = `(0|[1-9]\d*)`
+
+	preReleaseIdentifier = `(0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)`
+
+	buildIdentifier = `[0-9a-zA-Z-]+`
+
+	semverRe = regexp.MustCompile(`^` +
+
+		// dot-separated version segments (e.g. 1.2.3)
+		numericIdentifier + `\.` + numericIdentifier + `\.` + numericIdentifier +
+
+		// optional dot-separated prerelease segments (e.g. -alpha.PRERELEASE.1)
+		`(-` + preReleaseIdentifier + `(\.` + preReleaseIdentifier + `)*)?` +
```

Comment:
does semver require pre-release BEFORE build info?

Reply:
[Yes](https://semver.org/#backusnaur-form-grammar-for-valid-semver-versions).

I had started with a 1:1 copy of the [regexp from the spec](https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string) and then Jordan suggested this more readable variant.